lib/developer_portal/app/controllers/developer_portal/admin/account/invoices_controller.rb

```diff
@@ -1,25 +1,51 @@
+# frozen_string_literal: true
+
 class DeveloperPortal::Admin::Account::InvoicesController < ::DeveloperPortal::BaseController
+  class PaymentGatewayNotSupported < StandardError; end
+
+  rescue_from PaymentGatewayNotSupported, with: :handle_not_found
 
   helper Finance::InvoicesHelper
   helper Accounts::InvoicesHelper
 
   before_action :authorize_finance
   before_action :find_provider
+  before_action :authorize_payment_gateway, only: %i[payment payment_succeeded]
+
   activate_menu :account, :invoices
 
   liquify prefix: 'invoices'
 
   def index
-    collection = current_account.invoices.visible_for_buyer.page(params[:page])
-
+    collection = accessible_invoices.page(params[:page])
     invoices = Liquid::Drops::Invoice.wrap(collection)
     pagination = Liquid::Drops::Pagination.new(collection, self)
     assign_drops invoices: invoices, pagination: pagination
   end
 
   def show
-    invoice = current_account.invoices.visible_for_buyer.find(params[:id])
-    assign_drops invoice: Liquid::Drops::Invoice.wrap(invoice)
+    assign_drops invoice: Liquid::Drops::Invoice.wrap(find_invoice)
+  end
+
+  def payment
+    @invoice = find_invoice
+    payment_intent = @invoice.payment_intents.pending.latest.first!
```

Comment:
Can't we create the payment intent here if we do not find any?

Reply:
We can. In another PR though, I'd say.

Reply:
I think you do not want to create payment intents every time you display the payment page …

OK here should be the flow and I see the struggle because it is more complicated:

- On click on the “Pay” button
- send an AJAX request to create or retrieve a payment intent for this invoice
- get the response with the payment intent secret (https://stripe.com/docs/api/errors#errors-payment_intent-client_secret) and the publishable key and return them in the response
- use those keys in your stripe JS configuration
- submit Stripe Elements
- receive Stripe response and update the fields in the form
- submit the form to `InvoicesController#payment_succeeded`

Reply:
(I took the liberty to number your bullets above.)

For step 2, we need to be 100% certain not to end up confirming the payment intent then, just create it (in case it cannot be retrieved) without auto confirmation.

Reply:
Once it is released we cannot really do another PR because it involves changing customer front end again

Reply:
I disagree. The customer can remove the `pay_now?` check _if_ they want to, and actually there are good reasons not to as well.

Reply:
OK then :)